### PR TITLE
[AMDGPU][RFC] Combine asm and disasm tests.

### PR DESCRIPTION
**Pull request by kosarev**
Eliminates the need to replicate the same instructions in MC and MC/Disassembler tests and synchronize changes in them. Also highlights differences between disassembled, reassembled and original instructions.

### CODE REVIEW

**Comment by kosarev** [comment]
Another attempt, after https://github.com/llvm/llvm-project/pull/90179 and https://github.com/llvm/llvm-project/pull/90214.

We seem to use sed a lot in tests and I guess could even have a lit substitution, locally for AMDGPU or otherwise.

**Comment by ScottEgerton** [comment]
I love the idea behind this. However, the use of sed means this can no longer run on Windows.  I don’t think this is ideal personally. In any case, these tests would need to be constrained with “UNSUPPORTED: system-windows”.

**Comment by ScottEgerton** [comment]
I didn’t realise sed had been ported to Windows. Is there a plan to expand this to more tests? It looks good to me but I think it would be worthwhile to wait and see what the original reviewers of [#90179](https://github.com/llvm/llvm-project/pull/90179) and https://github.com/llvm/llvm-project/pull/90214 think.

**Comment by kosarev** [comment]
> Is there a plan to expand this to more tests?

Yes, that's very much the plan. The hope with this patch is that we could start having combined tests as we discuss alternative approaches involving changes in the common LLVM code.

Ping.

File: llvm/test/MC/AMDGPU/gfx12_asm_vop1.s [hunk 1]
```diff
@@ -1,3590 +1,3597 @@
-// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -show-encoding %s | FileCheck --check-prefix=GFX12 %s
-// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=-wavefrontsize32,+wavefrontsize64 -show-encoding %s | FileCheck --check-prefix=GFX12 %s
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -show-encoding %s | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-ASM %s
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -show-encoding %s | sed -n 's/.*encoding: \(\[.*\]\).*/\1/p' | llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -disassemble -show-encoding | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-DIS %s
```
Comment (arsenm) [inline llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:2]: This version does increase the run line complexity quite a bit. We have to maintain this regex in every test, which isn't great.

Comment (kosarev) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop1.s]: > Does `s/.*encoding://p` work?

Yes!

> This version does increase the run line complexity quite a bit. We have to maintain this regex in every test, which isn't great.

Added a lit substitution to address that.

**Comment by kosarev** [comment]
Ping.

**Comment by kosarev** [comment]
Ping.

**Comment by kosarev** [comment]
Ping. Can we land this?

**Comment by kosarev** [comment]
@arsenm Ping.

File: llvm/test/MC/AMDGPU/gfx12_asm_vop1.s [hunk 2]
```diff
@@ -1,3590 +1,3597 @@
...
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -show-encoding %s | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-ASM %s
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -show-encoding %s | %extract-encodings | llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -disassemble -show-encoding | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-DIS %s
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=-wavefrontsize32,+wavefrontsize64 -show-encoding %s | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-ASM %s
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=-wavefrontsize32,+wavefrontsize64 -show-encoding %s | %extract-encodings | llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=-wavefrontsize32,+wavefrontsize64 -disassemble -show-encoding | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-DIS %s
 
 v_bfrev_b32_e32 v5, v1
-// GFX12: encoding: [0x01,0x71,0x0a,0x7e]
+// GFX12: v_bfrev_b32_e32 v5, v1                  ; encoding: [0x01,0x71,0x0a,0x7e]
```
Comment (arsenm) [inline llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:7]: All the whitespace padding before the encoding is a bit annoying

File: llvm/test/MC/AMDGPU/gfx12_asm_vop1.s [hunk 3]
```diff
@@ -1,3590 +1,3597 @@
-// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -show-encoding %s | FileCheck --check-prefix=GFX12 %s
-// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=-wavefrontsize32,+wavefrontsize64 -show-encoding %s | FileCheck --check-prefix=GFX12 %s
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -show-encoding %s | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-ASM %s
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -show-encoding %s | %extract-encodings | llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,-wavefrontsize64 -disassemble -show-encoding | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-DIS %s
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=-wavefrontsize32,+wavefrontsize64 -show-encoding %s | FileCheck --strict-whitespace --check-prefixes=GFX12,GFX12-ASM %s
```
Comment (arsenm) [inline llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:3]: Don't need the -wavefrontsize32s

**Review by arsenm — APPROVED** [review]

**Review by Sisyph — APPROVED** [review]
LGTM too

File: llvm/test/MC/AMDGPU/gfx12_asm_vop1.s (hunk 3 above)
Comment (kosarev) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop1.s]: Removed via rebase.

File: llvm/test/MC/AMDGPU/gfx12_asm_vop1.s (hunk 2 above)
Comment (kosarev) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:7]: @MaskRay, would you mind having a command-line option to drop the padding?

Comment (MaskRay) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:7]: FileCheck treats consecutive spaces as one space by default. I don't think another option is needed.

Comment (kosarev) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:7]: We do `--strict-whitespace` to check spacing in disassembled instructions.

Comment (MaskRay) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:7]: OK, then I think `getCommentColumn` is fine to customize.

(But you merged this PR before the conversation is over. Now the test needs to be modified again.)

**Pull request merged**